cmd/cronosd/cmd/fix-unlucky-tx.go

```diff
@@ -0,0 +1,223 @@
+	evmtypes "github.com/tharsis/ethermint/x/evm/types"
+)
+
+const (
+	FlagMinBlockHeight = "min-block-height"
+
+	ExceedBlockGasLimitError = "out of gas in location: block gas meter; gasWanted:"
+)
+
+// FixUnluckyTxCmd update the tx execution result of false-failed tx in tendermint db
+func FixUnluckyTxCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "fix-unlucky-tx [blocks-file]",
+		Short: "Fix tx execution result of false-failed tx after v0.7.0 upgrade, \"-\" means stdin.",
+		Long:  "Fix tx execution result of false-failed tx after v0.7.0 upgrade.\nWARNING: don't use this command to patch blocks generated before v0.7.0 upgrade",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := server.GetServerContextFromCmd(cmd)
+			clientCtx := client.GetClientContextFromCmd(cmd)
+
+			minBlockHeight, err := cmd.Flags().GetInt(FlagMinBlockHeight)
+			if err != nil {
+				return err
+			}
+
+			chainID, err := cmd.Flags().GetString(flags.FlagChainID)
+			if err != nil {
+				return err
+			}
+
+			var blocksFile io.Reader
+			if args[0] == "-" {
+				blocksFile = os.Stdin
+			} else {
+				fp, err := os.Open(args[0])
+				if err != nil {
+					return err
+				}
+				defer fp.Close()
+				blocksFile = fp
+			}
+
+			scanner := bufio.NewScanner(blocksFile)
+			for scanner.Scan() {
+				blockNumber, err := strconv.ParseInt(scanner.Text(), 10, 64)
+				if err != nil {
+					return err
+				}
+
+				if blockNumber < int64(minBlockHeight) {
+					return fmt.Errorf("block number is generated before v0.7.0 upgrade: %d", blockNumber)
+				}
+
+				tmDB, err := openTMDB(ctx.Config, chainID)
+				if err != nil {
+					return err
+				}
+
+				// load results
+				blockResults, err := tmDB.stateStore.LoadABCIResponses(blockNumber)
+				if err != nil {
+					return err
+				}
+
+				// find unlucky tx
+				var txIndex int64
+				for i, txResult := range blockResults.DeliverTxs {
+					if TxExceedsBlockGasLimit(txResult) {
+						if len(txResult.Events) > 0 && txResult.Events[len(txResult.Events)-1].Type == evmtypes.TypeMsgEthereumTx {
+							// already patched
+							break
+						}
+
+						// load raw tx
+						blk := tmDB.blockStore.LoadBlock(blockNumber)
+						if blk == nil {
+							return fmt.Errorf("block not found: %d", blockNumber)
+						}
+
+						tx, err := clientCtx.TxConfig.TxDecoder()(blk.Txs[i])
+						if err != nil {
+							return err
+						}
+
+						txIndex++
+						for msgIndex, msg := range tx.GetMsgs() {
+							ethTxIndex := txIndex + int64(msgIndex)
+							ethTx, ok := msg.(*evmtypes.MsgEthereumTx)
+							if !ok {
+								continue
+							}
+							evt := abci.Event{
+								Type: evmtypes.TypeMsgEthereumTx,
+								Attributes: []abci.EventAttribute{
+									{Key: []byte(evmtypes.AttributeKeyEthereumTxHash), Value: []byte(ethTx.Hash), Index: true},
+									{Key: []byte(evmtypes.AttributeKeyTxIndex), Value: []byte(strconv.FormatInt(ethTxIndex, 10)), Index: true},
```

Comment:
This event format emulates the event format the tx is supposed to emit after this PR: https://github.com/tharsis/ethermint/pull/1062

Reply:
one thing I would like to clarify. By adding extra events, we are not modifying the block hash? 

Are the events taken out from the merkle tree computation? (I remember they were removed or added at some point)

Reply:
https://github.com/tendermint/tendermint/blob/044b12585ffde860fa02333b449d5ef7f1e4da7d/types/results.go#L47
yes, only these four fields are considered as deterministic.